### DISCUSSION TITLE
Web vs Android

### DISCUSSION THREAD

**Discussion by wz2b** [G-Two/subarulink#100] [General]
I'm curious, why go through the effort of reverse engineering this in Android, rather than just a web browser - is all the info in the app available on the desktop site as well, or no?

**Comment by G-Two** [comment]
Good question! When I first prototyped this almost 5 years ago, I did use the website API, and it worked fine. I ended up switching to the mobile app API because the JSON response at login was easier to work with compared to the HTML response from the website for things like enumerating multiple vehicles and recognizing if 2FA needed to be accomplished. It later ended being a matter of preference; unzipping an .apk every new app release and checking for API changes doesn't much more effort compared to gathering the same info via a web browser. New mobile app release notifications are an advantage since they prompt me that something has changed. With the web API, they'll just serve a new batch of .js files to implement the change. The disadvantage of using the mobile app API is that I do need to update the API endpoint version numbers (e.g. "g2v29") to keep up with new releases. The web API seems like it has been more or less constant, but there is no guarantee. With the mobile app API at least I do know "g2v29" should yield the same results until a new revision. I have seen Subaru change field names and drop fields between versions (which probably just happen without warning through the web API).

**Reply by wz2b** [reply]
Interesting.  What can you tell by unpacking the apk, though?  Is there more than jars or binaries in there?  I was thinking you were using a wedge of some sort to intercept traffic, not figuring it out from the binary.

**Comment by G-Two** [comment]
I did run it in an emulator at first (and intercepted traffic) to get a sense of how it behaved compared to the website. Then I saw that it was a Cordova app, with most of the useful functionality implemented in javascript and html files stored in the apk (within `assets/www/`. I never poked at any binary or native components of the apk.

**Reply by wz2b** [reply]
LOL That's awesome, we use Cordova at work!
